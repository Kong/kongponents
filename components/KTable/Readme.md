```js
const options = {
  headers: [
    { label: 'Name', key: 'name', sortable: true },
    { label: 'Type', key: 'type', sortable: true },
    { label: 'Credentials', key: 'credentials' },
    { label: '', key: 'actions', hideLabel: true }
  ],
  data: [
    {
      name: 'iOS App',
      type: 'Key',
      credentials: 'xyz'
    },
    {
      name: 'Website Desktop',
      type: 'Basic',
      credentials: '123xyz'
    },
    {
      name: 'Android App',
      type: 'JWT',
      credentials: '123xyz'
    }
  ]
}

<KTable :options=options :isStriped='true' :hasHover='true'>
  <template slot="actions" slot-scope="{row, rowKey, rowValue}"><a href="">Edit</a></template>
</KTable>
```
