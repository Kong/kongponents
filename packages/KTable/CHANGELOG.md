# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.3.1](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.3.0...@kongponents/ktable@0.3.1) (2020-04-10)


### Bug Fixes

* **ktable:** beef up defaultSorter to handle null, undefined and array values ([#206](https://github.com/Kong/kongponents/issues/206)) ([0cd8c76](https://github.com/Kong/kongponents/commit/0cd8c76c42e4a1c12209b5a041415e963360c905))





# [0.3.0](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.2.0...@kongponents/ktable@0.3.0) (2020-03-27)


### Features

* **ktable:** add row and cell events ([#205](https://github.com/Kong/kongponents/issues/205)) ([bb7ac1b](https://github.com/Kong/kongponents/commit/bb7ac1bcf70ef29d0c8df264fcfbfbd2b0697e49))





# [0.2.0](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.7...@kongponents/ktable@0.2.0) (2020-03-16)


### Features

* **ktable:** add sorting for ktable - fti-1291 ([#197](https://github.com/Kong/kongponents/issues/197)) ([19603d4](https://github.com/Kong/kongponents/commit/19603d4d5f7340016ab6c8c7158b12d3019abe49))





## [0.1.7](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.6...@kongponents/ktable@0.1.7) (2020-03-06)

**Note:** Version bump only for package @kongponents/ktable





## [0.1.6](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.4-7...@kongponents/ktable@0.1.6) (2020-01-16)

**Note:** Version bump only for package @kongponents/ktable





## [0.1.4-7](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.4-6...@kongponents/ktable@0.1.4-7) (2019-12-17)

**Note:** Version bump only for package @kongponents/ktable





## [0.1.4-6](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.4-5...@kongponents/ktable@0.1.4-6) (2019-12-16)

**Note:** Version bump only for package @kongponents/ktable





## [0.1.4-5](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.4-4...@kongponents/ktable@0.1.4-5) (2019-11-12)

**Note:** Version bump only for package @kongponents/ktable





## [0.1.4-4](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.4-3...@kongponents/ktable@0.1.4-4) (2019-11-11)

**Note:** Version bump only for package @kongponents/ktable





## [0.1.4-3](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.4-2...@kongponents/ktable@0.1.4-3) (2019-11-08)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))





## [0.1.4-2](https://github.com/Kong/kongponents/compare/@kongponents/ktable@0.1.4-1...@kongponents/ktable@0.1.4-2) (2019-11-08)

**Note:** Version bump only for package @kongponents/ktable





# @kongponents/ktable

## Table of Contents

- [0.1.4](#014)

### 0.1.4
- Bugfix, remove extra whitespace around row keys [#112](https://github.com/Kong/kongponents/pull/112)
