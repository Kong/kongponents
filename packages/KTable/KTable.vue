<template>
  <table
    :class="{'has-hover': hasHover, 'is-small': isSmall}"
    class="k-table">
    <thead>
      <tr>
        <template>
          <th
            v-for="(column, index) in options.headers"
            :key="index">
            <slot
              v-if="!column.hideLabel"
              :name="`column-${column.key}`"
              :column="column">
              {{ column.label ? column.label : column.key }}
            </slot>
          </th>
        </template>
      </tr>
    </thead>
    <tbody>
      <tr
        v-for="(row, rowIndex) in options.data"
        :key="rowIndex">
        <template>
          <td
            v-for="(value, index) in options.headers"
            :key="index">
            <slot
              :name="value.key"
              :row="row"
              :rowKey="rowIndex"
              :rowValue="row[value.key]">
              {{ row[value.key] }}
            </slot>
          </td>
        </template>
      </tr>
    </tbody>
  </table>
</template>

<script>
export default {
  name: 'KTable',
  props: {
    /**
     * Object containing data which creates rows and columns.
     * @param {Object} options - Options to initialize the component with
     * @param {Array} options.headers - Array of Objects defining Table Headers
     * @param {Array} options.data - Array of Objects defining column data
     */
    options: {
      type: Object,
      required: true
    },
    /**
     * Enables hover highlighting to table rows
     */
    hasHover: {
      type: Boolean,
      default: false
    },
    /**
     * Lowers overall table padding
     */
    isSmall: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style scoped lang="scss">
@import '~@kongponents/styles/_variables.scss';

table.k-table {
  border-collapse: collapse;
}

.k-table {
  width: 100%;
  max-width: 100%;
  th,
  td {
    padding: var(--spacing-md, spacing(md));
    vertical-align: middle;
  }
  thead {
    border-top: 1px solid var(--KTableBorder, var(--grey92, color(grey-92)));
    border-bottom: 2px solid var(--KTableBorder, var(--grey92, color(grey-92)));
    th {
      padding: var(--spacing-sm, spacing(sm)) var(--spacing-md, spacing(md));
      text-align: left;
      font-size: var(--KTableHeaderSize, var(--type-sm, type(sm)));
      font-weight: 500;
    }
  }
  tbody {
    tr {
      border-bottom: 1px solid var(--KTableBorder, var(--grey92, color(grey-92)));
    }
    td {
      color: var(--KTableColor, var(--tblack-70, color(tblack-70)));
      a {
        color: var(--blue-link, color(blue-link));
        text-decoration: none;
        &:hover {
          text-decoration: underline;
        }
      }
    }
  }

  // Variants
  &.is-small {
    th {
      padding: var(--spacing-xs, spacing(xs)) var(--spacing-sm, spacing(sm));
    }
    td {
      padding: var(--spacing-sm, spacing(sm));
    }
  }
  &.has-hover tbody tr:hover {
    background-color: var(--KTableHover, var(--blue-lightest, color(blue-lightest)));
  }
}

</style>
