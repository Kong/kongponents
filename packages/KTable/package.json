{
  "name": "@kongponents/ktable",
  "version": "7.1.4",
  "description": "Table component",
  "main": "dist/KTable.umd.min.js",
  "componentName": "KTable",
  "source": "KTable.vue",
  "files": [
    "dist",
    "KTable.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kbutton": "^7.1.4",
    "@kongponents/kemptystate": "^7.1.4",
    "@kongponents/kicon": "^7.1.4",
    "@kongponents/kpagination": "^7.1.4",
    "@kongponents/kskeleton": "^7.1.4",
    "@kongponents/styles": "^7.1.4",
    "@kongponents/utils": "^7.1.4",
    "vue-uuid": "^2.0.2"
  },
  "peerDependencies": {
    "vue": ">= 2.7.0 < 3"
  }
}
