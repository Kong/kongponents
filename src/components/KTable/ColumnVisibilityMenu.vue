<template>
  <div class="table-column-visibility-menu">
    <KDropdown
      data-testid="table-column-visibility-menu"
      @toggle-dropdown="handleDropdownToggle"
    >
      <KTooltip text="Show/Hide Columns">
        <KButton
          appearance="tertiary"
          class="menu-button"
          data-testid="column-visibility-menu-button"
        >
          <template #icon>
            <ViewColumns />
          </template>
        </KButton>
      </KTooltip>

      <template #items>
        <KDropdownItem
          v-for="col in columns"
          :key="col.key"
          class="column-visibility-menu-item"
          :data-testid="`column-visibility-menu-item-${col.key}`"
          @click.stop="() => {
            visibilityMap[col.key] = !visibilityMap[col.key]
            isDirty = true
          }"
        >
          <!-- KLabel must be separate to maintain click handling on the label within the dropdown item -->
          <KCheckbox
            v-model="visibilityMap[col.key]"
            :aria-labelledby="`${tableId}-${col.key}-visibility-checkbox-label`"
            :data-testid="`column-visibility-checkbox-${col.key}`"
          />
          <KLabel
            :id="`${tableId}-${col.key}-visibility-checkbox-label`"
            class="visibility-checkbox-label"
          >
            {{ col.label }}
          </KLabel>
        </KDropdownItem>
        <div class="apply-button-wrapper">
          <KButton
            appearance="tertiary"
            class="apply-button"
            data-testid="apply-button"
            @click="handleApply"
          >
            Apply
          </KButton>
        </div>
      </template>
    </KDropdown>
  </div>
</template>

<script setup lang="ts">
import { ref, watch, onBeforeMount, type PropType } from 'vue'
import type { TableHeader } from '@/types'
import { ViewColumns } from '@kong/icons'
import KButton from '@/components/KButton/KButton.vue'
import KCheckbox from '@/components/KCheckbox/KCheckbox.vue'
import KDropdown from '@/components/KDropdown/KDropdown.vue'
import KDropdownItem from '@/components/KDropdown/KDropdownItem.vue'

const emit = defineEmits<{
  (e: 'update:visibility', columnVisibility: Record<string, boolean>): void
}>()

const props = defineProps({
  columns: {
    type: Array as PropType<TableHeader[]>,
    required: true,
  },
  tableId: {
    type: String,
    required: true,
  },
  visibilityPreferences: {
    type: Object as PropType<Record<string, boolean>>,
    default: () => ({}),
  },
})

const visibilityMap = ref<Record<string, boolean>>({})
const isDirty = ref(false)

const initVisibilityMap = (): void => {
  visibilityMap.value = props.columns.reduce((acc, col: TableHeader) => {
    acc[col.key] = props.visibilityPreferences[col.key] === undefined ? true : props.visibilityPreferences[col.key]

    return acc
  }, {} as Record<string, boolean>)
  isDirty.value = false
}

const handleApply = (): void => {
  // pass by ref problems
  emit('update:visibility', JSON.parse(JSON.stringify(visibilityMap.value)))
  isDirty.value = false
}

const handleDropdownToggle = (isOpen: boolean): void => {
  // reset the map if the dropdown is closed without applying changes
  if (!isOpen && isDirty.value) {
    initVisibilityMap()
  }
}

watch(() => props.visibilityPreferences, () => {
  initVisibilityMap()
})

onBeforeMount(() => {
  // initialize visibility state
  initVisibilityMap()
})
</script>

<style lang="scss" scoped>
.table-column-visibility-menu {
  margin-left: var(--kui-space-auto, $kui-space-auto);

  .apply-button-wrapper {
    display: flex;
    width: 100%;

    .apply-button {
      margin-left: var(--kui-space-auto, $kui-space-auto);
      margin-right: var(--kui-space-auto, $kui-space-auto);
    }
  }

  .visibility-checkbox-label {
    cursor: pointer;
    margin-bottom: var(--kui-space-0, $kui-space-0);
    margin-left: calc(-1 * var(--kui-space-40, $kui-space-40)); // because dropdown item container and checkbox both have default spacing, reduce it
  }
}
</style>
