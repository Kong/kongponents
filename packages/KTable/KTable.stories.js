import Vue from 'vue'
import { storiesOf } from '../../storybook/node_modules/@storybook/vue'

import KTable from './KTable.vue'

const options = {
  headers: [
    { label: 'Name', key: 'name' },
    { label: 'ID', key: 'id' },
    { label: 'Enabled', key: 'enabled' },
    { key: 'actions', hideLabel: true }
  ],
  data: [
    {
      name: 'Basic Auth',
      id: '517526354743085',
      enabled: 'true'
    },
    {
      name: 'Website Desktop',
      id: '328027447731198',
      enabled: 'false'
    },
    {
      name: 'Android App',
      id: '405383051040955',
      enabled: 'true'
    }
  ]
}

storiesOf('Table', module)
  .add('Default', () => ({
    components: { KTable },
    template: `<KTable :options=options />`,
    data () {
      return {
        options
      }
    }
  }), {info: {}})
  .add('hover', () => ({
    components: { KTable },
    template: `<KTable :options=options :hasHover='true' />`,
    data () {
      return {
        options
      }
    }
  }), {info: {}})
  .add('small', () => ({
    components: { KTable },
    template: `<KTable :options="options" is-small />`,
    data () {
      return {
        options
      }
    }
  }), {info: {}})
  .add('slots', () => ({
    components: { KTable },
    template: `
    <KTable :options=options>
      <template slot="name" slot-scope="{row, rowKey, rowValue}">
        <strong v-if="rowValue === 'Basic Auth'">{{ rowValue }}</strong>
      </template>
      <template slot="enabled" slot-scope={rowValue}>
        <span v-if="rowValue === 'true'" style="color: green">&#10003;</span>
        <span v-else style="color: red">&#10007;</span>
      </template>
      <template slot="actions"><a href="">Edit</a></template>
    </KTable>
    <!--
    data() {
      return {
        options: ${JSON.stringify(options, null, 2)}
      }
    }
    -->
    `,
    data () {
      return {
        options
      }
    }
  }), {info: {}})
