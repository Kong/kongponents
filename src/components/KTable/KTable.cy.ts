import { mount } from 'cypress/vue'
import { h } from 'vue'
import KTable from '@/components/KTable/KTable.vue'

const largeDataSet = [
  {
    name: 'Basic Auth',
    id: '517526354743085',
    enabled: 'true',
  },
  {
    name: 'Website Desktop',
    id: '328027447731198',
    enabled: 'false',
  },
  {
    name: 'Android App',
    id: '405383051040955',
    enabled: 'true',
  },
  {
    name: 'Basic Auth',
    id: '517526354743085',
    enabled: 'true',
  },
  {
    name: 'Website Desktop',
    id: '328027447731198',
    enabled: 'false',
  },
  {
    name: 'Android App',
    id: '405383051040955',
    enabled: 'true',
  },
  {
    name: 'Basic Auth',
    id: '517526354743085',
    enabled: 'true',
  },
  {
    name: 'Website Desktop',
    id: '328027447731198',
    enabled: 'false',
  },
  {
    name: 'Android App',
    id: '405383051040955',
    enabled: 'true',
  },
  {
    name: 'Basic Auth',
    id: '517526354743085',
    enabled: 'true',
  },
  {
    name: 'Website Desktop',
    id: '328027447731198',
    enabled: 'false',
  },
  {
    name: 'Android App',
    id: '405383051040955',
    enabled: 'true',
  },
]

const options = {
  headers: [
    { label: 'Name', key: 'name', sortable: true, hideLabel: false },
    { label: 'ID', key: 'id', sortable: false, hideLabel: false },
    { label: 'Enabled', key: 'enabled', sortable: false, hideLabel: false },
    { label: '', key: 'actions', sortable: false, hideLabel: true },
  ],
  data: [
    {
      name: 'Basic Auth',
      id: '517526354743085',
      enabled: 'true',
    },
    {
      name: 'Website Desktop',
      id: '328027447731198',
      enabled: 'false',
    },
    {
      name: 'Android App',
      id: '405383051040955',
      enabled: 'true',
    },
  ],
}

/**
 * ALL TESTS MUST USE testMode
 * We generate unique IDs for reference by aria properties. Test mode strips these out
 * allowing for successful snapshot verification.
 * props: {
 *   testMode: 'true' || 'loading'
 * }
 */

describe('KTable', () => {
  describe('default', () => {
    it('renders link in action slot', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          headers: options.headers,
          fetcher: () => { return { data: options.data } },
          disablePagination: true,
        },
        slots: {
          actions: () => h('a', { href: '#' }, 'Link'),
        },
      })

      cy.get('.k-table td:last-of-type > *').contains('a', 'Link')
    })

    it('renders content in the toolbar slot', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          headers: options.headers,
          fetcher: () => { return { data: options.data } },
          disablePagination: true,
        },
        slots: {
          toolbar: () => h('button', {}, 'Toolbar button'),
        },
      })

      cy.get('.k-table-container .k-table-toolbar').find('button').should('be.visible')
      cy.get('.k-table-container .k-table-toolbar button').should('contain.text', 'Toolbar button')
    })

    it('has hover class when passed', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          headers: options.headers,
          fetcher: () => { return { data: options.data } },
          hasHover: true,
        },
      })

      cy.get('.k-table').should('have.class', 'has-hover')
    })
  })

  describe('sorting', () => {
    it('should have sortable class when passed', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          headers: options.headers,
          fetcher: () => { return { data: options.data } },
        },
      })

      cy.get('th').each(($el, index) => {
        if (index === 0) {
          cy.wrap($el).should('have.class', 'sortable')
        }
      })
    })

    it('should allow disabling sorting', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          headers: options.headers,
          fetcher: () => { return { data: options.data } },
          disableSorting: true,
        },
      })

      cy.get('th').each(($el) => {
        cy.wrap($el).should('not.have.class', 'sortable')
      })
    })
  })

  // describe('events', () => {
  //   it('@row:event', () => {
  //     const evtTrigger = jest.fn()
  //     mount(KTable, {
  //       localVue,
  //       attachToDocument: true,
  //       props: {
  //         testMode: 'true',
  //         headers: options.headers,
  //         fetcher: () => { return { data: options.data } },
  //       },
  //       listeners: {
  //         'row:mouseover': evtTrigger,
  //       },
  //     })
  //     const bodyRow = wrapper.find('.k-table tbody tr')
  //     bodyRow.trigger('mouseover')
  //     expect(evtTrigger).toHaveBeenNthCalledWith(1, expect.objectContaining({ type: 'mouseover' }), options.data[0], 'row')
  //   })

  // it('@cell:event', () => {
  //   const evtTrigger = jest.fn()
  //   mount(KTable, {
  //     localVue,
  //     attachToDocument: true,
  //     props: {
  //       testMode: 'true',
  //       headers: options.headers,
  //       fetcher: () => { return { data: options.data } },
  //     },
  //     listeners: {
  //       'cell:click': evtTrigger,
  //       'cell:mouseover': evtTrigger,
  //       'cell:mouseout': evtTrigger,
  //     },
  //   })

  //

  //     const bodyCell1 = wrapper.find('.k-table tbody td')
  //     const bodyCell2 = wrapper.find('.k-table tbody td:nth-child(2)')

  //     bodyCell1.trigger('click')
  //     bodyCell1.trigger('mouseover')
  //     bodyCell2.trigger('mouseout')
  //     expect(evtTrigger).toHaveBeenNthCalledWith(1, expect.objectContaining({ type: 'click' }), 'Basic Auth', 'cell')
  //     expect(evtTrigger).toHaveBeenNthCalledWith(2, expect.objectContaining({ type: 'mouseover' }), 'Basic Auth', 'cell')
  //     expect(evtTrigger).toHaveBeenNthCalledWith(3, expect.objectContaining({ type: 'mouseout' }), '517526354743085', 'cell')
  //   })
  // })

  describe('states', () => {
    it('displays an empty state when no data is available', () => {
      const fetcher = () => new Promise(resolve => resolve({ data: [] }))
      mount(KTable, {
        props: {
          testMode: 'true',
          fetcher,
          headers: options.headers,
          pageSize: 4,
        },
      })

      cy.get('.empty-state-wrapper').should('be.visible')
    })

    it('displays an empty state when no data is available (slot)', () => {
      const emptySlotContent = 'Look mah! I am empty! (except testMode)'
      const fetcher = () => new Promise(resolve => resolve({ data: [] }))
      mount(KTable, {
        props: {
          testMode: 'true',
          fetcher,
          headers: options.headers,
          pageSize: 4,
        },
        slots: {
          'empty-state': () => h('span', {}, emptySlotContent),
        },
      })

      cy.getTestId('k-table-empty-state').should('contain.text', emptySlotContent)
    })

    it('displays a loading skeletion when the "isLoading" prop is set to true"', () => {
      mount(KTable, {
        props: {
          testMode: 'loading',
          isLoading: true,
        },
      })

      cy.get('.skeleton-table-wrapper').should('be.visible')
    })

    it('displays an error state when the "hasError" prop is set to true"', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          hasError: true,
        },
      })

      cy.get('.empty-state-wrapper').should('be.visible')
      cy.get('.is-error').should('be.visible')
    })

    it('displays an error state (slot)', () => {
      const errorSlotContent = 'Look mah! I am erroneous! (except testMode)'
      mount(KTable, {
        props: {
          testMode: 'true',
          hasError: true,
        },
        slots: {
          'error-state': () => h('span', {}, errorSlotContent),
        },
      })

      cy.getTestId('k-table-error-state').should('contain.text', errorSlotContent)
    })

    it('displays a loading state and not an empty state when pending response', () => {
      const slowFetcher = () => {
        return new Promise((resolve) => setTimeout(resolve, 2500))
      }

      mount(KTable, {
        props: {
          testMode: 'loading',
          fetcher: slowFetcher,
          headers: options.headers,
          paginationPageSizes: [10, 20, 30, 40],
        },
      })

      cy.get('.skeleton-table-wrapper').should('be.visible')
      cy.get('.empty-state-wrapper').should('not.exist')
    })
  })

  describe('pagination', () => {
    it('displays pagination when fetcher provided', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          fetcher: () => {
            return largeDataSet
          },
          isLoading: false,
          headers: options.headers,
          paginationPageSizes: [10, 20, 30, 40],
        },
      })

      cy.getTestId('k-table-pagination').should('be.visible')
    })

    it('does not display pagination when pagination disabled', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          fetcher: () => {
            return largeDataSet
          },
          isLoading: false,
          headers: options.headers,
          paginationPageSizes: [10, 20, 30, 40],
          disablePagination: true,
        },
      })

      cy.getTestId('k-table-pagination').should('not.exist')
    })

    it('does not display pagination when no fetcher', () => {
      mount(KTable, {
        props: {
          testMode: 'true',
          options,
          paginationPageSizes: [10, 20, 30, 40],
        },
      })

      cy.getTestId('k-table-pagination').should('not.exist')
    })

    it('does not display pagination when hidePaginationWhenOptional is true and total is less than min pageSize', () => {
      mount(KTable, {
        propsData: {
          testMode: 'true',
          fetcher: () => { return { data: options.data, total: options.data.length } },
          isLoading: false,
          headers: options.headers,
          paginationPageSizes: [10, 15, 20],
          hidePaginationWhenOptional: true,
        },
      })

      cy.getTestId('k-table-pagination').should('not.exist')
    })

    it('does not display pagination when hidePaginationWhenOptional is true and total is equal to min pageSize', () => {
      mount(KTable, {
        propsData: {
          testMode: 'true',
          fetcher: () => {
            return { data: largeDataSet, total: 12 }
          },
          isLoading: false,
          headers: options.headers,
          paginationPageSizes: [12, 15, 20],
          hidePaginationWhenOptional: true,
        },
      })

      cy.getTestId('k-table-pagination').should('not.exist')
    })

    it('does display pagination when total is greater than min pageSize', () => {
      mount(KTable, {
        propsData: {
          testMode: 'true',
          fetcher: () => {
            return { data: largeDataSet, total: 12 }
          },
          isLoading: false,
          headers: options.headers,
          paginationPageSizes: [10, 15, 20],
          hidePaginationWhenOptional: true,
        },
      })

      cy.getTestId('k-table-pagination').should('be.visible')
    })

    it('does not display offset-based pagination when hidePaginationWhenOptional is true and total is less than min pageSize', () => {
      mount(KTable, {
        propsData: {
          testMode: 'true',
          fetcher: () => { return { data: options.data, offset: null } },
          isLoading: false,
          headers: options.headers,
          paginationPageSizes: [10, 15, 20],
          hidePaginationWhenOptional: true,
          initialFetcherParams: { offset: null },
          paginationType: 'offset',
        },
      })

      cy.getTestId('k-table-pagination').should('not.exist')
    })

    it('does display offset-based pagination when total is greater than min pageSize', () => {
      mount(KTable, {
        propsData: {
          testMode: 'true',
          fetcher: () => {
            return { data: largeDataSet, offset: 'abc' }
          },
          isLoading: false,
          initialFetcherParams: { offset: 'abc' },
          headers: options.headers,
          paginationPageSizes: [10, 15, 20],
          hidePaginationWhenOptional: true,
          paginationType: 'offset',
        },
      })

      cy.getTestId('k-table-pagination').should('be.visible')
    })
  })

  describe('misc', () => {
    it('triggers the internal search and revalidate after clearing the search input', () => {
      const fns = {
        fetcher: ({ query }: { query: string }) => {
          return { data: [{ query }] }
        },
      }

      cy.spy(fns, 'fetcher').as('fetcher')

      mount(KTable, {
        propsData: {
          testMode: 'true',
          fetcher: fns.fetcher,
          isLoading: false,
          initialFetcherParams: { offset: 'abc' },
          headers: options.headers,
          paginationPageSizes: [10, 15, 20],
          hidePaginationWhenOptional: true,
          paginationType: 'offset',
          searchInput: '',
        },
      })
        .get('@fetcher')
        .should('have.callCount', 1) // fetcher's 1st call
        .should('returned', { data: [{ query: '' }] })
        .wait(1000)
        .get('@fetcher')
        .should('have.callCount', 1) // ensure fetcher is NOT called twice on load
        .then(() => cy.wrap(Cypress.vueWrapper.setProps({ searchInput: 'some-keyword' })))

      // fetcher call should be delayed (> 350ms for search func + 500ms for revalidate func)
      cy.get('@fetcher', { timeout: 1000 }) // fetcher's 2nd call
        .should('have.callCount', 2) // fetcher should be called once
        .should('returned', { data: [{ query: 'some-keyword' }] })
        .then(() => cy.wrap(Cypress.vueWrapper.setProps({ searchInput: '' })))

      // fetcher should be called immediately (< 350ms for search func)
      cy.get('@fetcher', { timeout: 350 })
        .should('have.callCount', 3) // fetcher's 3rd call
        .should('returned', { data: [{ query: '' }] })
    })
  })
})
