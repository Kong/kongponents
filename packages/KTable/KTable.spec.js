import { mount } from '@vue/test-utils'
import KTable from '@/KTable/KTable'

const options = {
  headers: [
    { label: 'Name', key: 'name' },
    { label: 'ID', key: 'id' },
    { label: 'Enabled', key: 'enabled' },
    { key: 'actions', hideLabel: true }
  ],
  data: [
    {
      name: 'Basic Auth',
      id: '517526354743085',
      enabled: 'true'
    },
    {
      name: 'Website Desktop',
      id: '328027447731198',
      enabled: 'false'
    },
    {
      name: 'Android App',
      id: '405383051040955',
      enabled: 'true'
    }
  ]
}

describe('KTable', () => {
  it('renders link in action slot', () => {
    const wrapper = mount(KTable, {
      propsData: {
        options
      },
      scopedSlots: {
        actions: '<a href="#" slot-scope="actions">Link</a>'
      }
    })

    const actions = wrapper.findAll('.k-table td:last-of-type > *')

    expect(actions.at(0).is('a')).toBe(true)
    expect(actions.at(1).is('a')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('has hover class when passed', () => {
    const wrapper = mount(KTable, {
      propsData: {
        options,
        hasHover: true
      }
    })

    expect(wrapper.classes()).toContain('has-hover')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('has small class when passed', () => {
    const wrapper = mount(KTable, {
      propsData: {
        options,
        isSmall: true
      }
    })

    expect(wrapper.classes()).toContain('is-small')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
