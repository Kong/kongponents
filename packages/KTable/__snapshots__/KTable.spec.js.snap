// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KTable has hover class when passed 1`] = `
<table class="k-table has-hover">
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        ID
      </th>
      <th>
        Enabled
      </th>
      <th>
        <!---->
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Basic Auth
      </td>
      <td>
        517526354743085
      </td>
      <td>
        true
      </td>
      <td>

      </td>
    </tr>
    <tr>
      <td>
        Website Desktop
      </td>
      <td>
        328027447731198
      </td>
      <td>
        false
      </td>
      <td>

      </td>
    </tr>
    <tr>
      <td>
        Android App
      </td>
      <td>
        405383051040955
      </td>
      <td>
        true
      </td>
      <td>

      </td>
    </tr>
  </tbody>
</table>
`;

exports[`KTable has small class when passed 1`] = `
<table class="k-table is-small">
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        ID
      </th>
      <th>
        Enabled
      </th>
      <th>
        <!---->
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Basic Auth
      </td>
      <td>
        517526354743085
      </td>
      <td>
        true
      </td>
      <td>

      </td>
    </tr>
    <tr>
      <td>
        Website Desktop
      </td>
      <td>
        328027447731198
      </td>
      <td>
        false
      </td>
      <td>

      </td>
    </tr>
    <tr>
      <td>
        Android App
      </td>
      <td>
        405383051040955
      </td>
      <td>
        true
      </td>
      <td>

      </td>
    </tr>
  </tbody>
</table>
`;

exports[`KTable renders link in action slot 1`] = `
<table class="k-table">
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        ID
      </th>
      <th>
        Enabled
      </th>
      <th>
        <!---->
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Basic Auth
      </td>
      <td>
        517526354743085
      </td>
      <td>
        true
      </td>
      <td><a href="#">Link</a></td>
    </tr>
    <tr>
      <td>
        Website Desktop
      </td>
      <td>
        328027447731198
      </td>
      <td>
        false
      </td>
      <td><a href="#">Link</a></td>
    </tr>
    <tr>
      <td>
        Android App
      </td>
      <td>
        405383051040955
      </td>
      <td>
        true
      </td>
      <td><a href="#">Link</a></td>
    </tr>
  </tbody>
</table>
`;
